coderd/cryptokeys/dbkeycache.go

```diff
@@ -0,0 +1,189 @@
+	now := d.clock.Now()
+	d.keysMu.RLock()
+	key, ok := d.keys[sequence]
+	d.keysMu.RUnlock()
+	if ok {
+		return checkKey(key, now)
+	}
+
+	d.keysMu.Lock()
+	defer d.keysMu.Unlock()
+
+	key, ok = d.keys[sequence]
+	if ok {
+		return checkKey(key, now)
+	}
+
+	err := d.fetch(ctx)
+	if err != nil {
+		return codersdk.CryptoKey{}, xerrors.Errorf("fetch: %w", err)
+	}
+
+	key, ok = d.keys[sequence]
+	if !ok {
+		return codersdk.CryptoKey{}, ErrKeyNotFound
+	}
+
+	return checkKey(key, now)
+}
+
+// Signing returns the latest valid key for signing. A valid key is one that is
+// both past its start time and before its deletion time.
+func (d *DBCache) Signing(ctx context.Context) (codersdk.CryptoKey, error) {
+	d.keysMu.RLock()
+	latest := d.latestKey
+	d.keysMu.RUnlock()
+
+	now := d.clock.Now()
+	if latest.CanSign(now) {
+		return db2sdk.CryptoKey(latest), nil
+	}
+
+	d.keysMu.Lock()
+	defer d.keysMu.Unlock()
+
+	if d.latestKey.CanSign(now) {
+		return db2sdk.CryptoKey(d.latestKey), nil
+	}
+
+	// Refetch all keys for this feature so we can find the latest valid key.
+	err := d.fetch(ctx)
+	if err != nil {
+		return codersdk.CryptoKey{}, xerrors.Errorf("fetch: %w", err)
+	}
+
+	return db2sdk.CryptoKey(d.latestKey), nil
+}
+
+// clear invalidates the cache. This forces the subsequent call to fetch fresh keys.
+func (d *DBCache) clear() {
+	now := d.clock.Now("DBCache", "clear")
+	d.keysMu.Lock()
+	defer d.keysMu.Unlock()
+	// Check if we raced with a fetch. It's possible that the timer fired and we
+	// lost the race to the mutex. We want to avoid invalidating
+	// a cache that was just refetched.
+	if now.Before(d.invalidateAt) {
+		return
+	}
+	d.keys = nil
+	d.latestKey = database.CryptoKey{}
+}
+
+// fetch fetches all keys for the given feature and determines the latest key.
+// It must be called while holding the keysMu lock.
+func (d *DBCache) fetch(ctx context.Context) error {
+	keys, err := d.db.GetCryptoKeysByFeature(ctx, d.feature)
+	if err != nil {
+		return xerrors.Errorf("get crypto keys by feature: %w", err)
+	}
+
+	now := d.clock.Now()
+	d.timer.Stop()
+	d.timer = d.newTimer()
+	d.invalidateAt = now.Add(time.Minute * 10)
+
+	cache := make(map[int32]database.CryptoKey)
+	var latest database.CryptoKey
+	for _, key := range keys {
+		cache[key.Sequence] = key
+		if key.CanSign(now) && key.Sequence > latest.Sequence {
+			latest = key
+		}
+	}
+
+	if len(cache) == 0 {
+		return ErrKeyNotFound
+	}
+
+	if !latest.CanSign(now) {
+		return ErrKeyInvalid
+	}
+
+	d.keys, d.latestKey = cache, latest
+	return nil
+}
+
+func checkKey(key database.CryptoKey, now time.Time) (codersdk.CryptoKey, error) {
+	if !key.CanVerify(now) {
+		return codersdk.CryptoKey{}, ErrKeyInvalid
+	}
+
+	return db2sdk.CryptoKey(key), nil
+}
+
+func (d *DBCache) newTimer() *quartz.Timer {
+	return d.clock.AfterFunc(time.Minute*10, d.clear)
+}
+
+func (d *DBCache) Close() {
+	d.keysMu.Lock()
+	defer d.keysMu.Unlock()
+	d.timer.Stop()
```

Comment:
So, it's good that you hold the mutex here, but if we need to be concerned with other things calling the `DBCache` at the same time (we do), then we also need to ensure another call doesn't come along after this one, trigger a fetch, and restart the timer.

You could add a `closed` field, or just nil out the timer.  The question is, when to check?  We could check on every call to Signing / Verifying --- basically refuse to answer any more key queries after we are closed.  

That seems a little unfriendly if things are getting closed down in multiple goroutines. It might be nice to serve keys we have locally, but refuse to query the database. The danger there is if we close the DBCache, but something else still has a reference to it, it could happily serve keys forever without erroring.

So, probably the safest thing is to stop answering key queries once we are closed.